examples/AvroGeneric/Program.cs

```diff
@@ -79,23 +79,28 @@ static void Main(string[] args)
             using (var consumer = new Consumer<string, GenericRecord>(consumerConfig, new AvroDeserializer<string>(), new AvroDeserializer<GenericRecord>()))
             using (var producer = new Producer<string, GenericRecord>(producerConfig, new AvroSerializer<string>(), new AvroSerializer<GenericRecord>()))
             {
-                consumer.OnMessage += (o, e)
-                    => Console.WriteLine($"Key: {e.Key}\nValue: {e.Value}");
-
                 consumer.OnError += (_, e)
                     => Console.WriteLine("Error: " + e.Reason);
 
-                consumer.OnConsumeError += (_, e)
-                    => Console.WriteLine("Consume error: " + e.Error.Reason);
-
                 consumer.Subscribe(topicName);
 
                 CancellationTokenSource cts = new CancellationTokenSource();
                 var consumeTask = Task.Factory.StartNew(() =>
                 {
                     while (!cts.Token.IsCancellationRequested)
                     {
-                        consumer.Poll(100);
+                        try
+                        {
+                            var consumeResult = consumer.Consume(TimeSpan.FromMilliseconds(100));
+                            if (consumeResult.Message != null)
+                            {
+                                Console.WriteLine($"Key: {consumeResult.Message.Key}\nValue: {consumeResult.Value}");
+                            }
+                        }
+                        catch (ConsumeException e)
+                        {
+                            Console.WriteLine("Consume error: " + e.Error.Reason);
+                        }
```

Comment:
Perhaps of token is added to `Task.Factory.StartNew()` we can omit line 124 and replace it with consumer.Close() ?

Reply:
@edenhill - `Close` is not in this branch yet - `Dispose` is blocking (and closes). That said, I did happen to forgot to add it in this location in the branch where I have made this change.

@kuskmen - thanks for chiming in! First, though you didn't note this, I've changed `Task.Factory.StartNew` to `Task.Run`. I think I probably used `StartNew` here due to copy/pasting from the producer poll thread (where it is appropriate). Second, I've added the token as a parameter to `Run` ( https://stackoverflow.com/questions/3712939/cancellation-token-in-task-constructor-why ). I have also replaced the `consumeTask.Wait()` with `consumer.Close` as you suggest. But I have an open question:

@edenhill - is it safe to call `rd_kafka_consumer_close`, or `rd_kafka_destroy` whist there is a blocking `rd_kafka_consumer_poll` executing in a separate thread (don't see this documented).

Reply:
clarification: `Close` is in fact in this branch (forgot about that), but it's not implemented as we want yet - is just a synonym for `Dispose`.

Reply:
@mhowlett You must not call destroy() on an rd_kafka_t instance that is referenced/used elsewhere, it must be the last thing that happens to that instance.   consumer_close is okay though

Reply:
Ok, makes sense. I think in .NET though we should add tracking / synchronization to ensure `Dispose` waits for any in-flight method calls before calling `rd_kafka_destroy` - we need to avoid the possibility of access violation exceptions since I'm sure people are going to write code that will cause them and they'll get a bad impression of the library if they do. I won't do this as a response to this PR though, I'll just add a todo (and make sure all our code is correct).